lib/blocking_queue.ex

```diff
@@ -144,6 +144,11 @@ defmodule BlockingQueue do
     {:reply, :queue.member(item, queue), {max, queue}}
   end
 
+  # remove all items using predicate function
+  def handle_call({:filter, f}, _, {max, queue}) do
+    {:reply, nil, {max, :queue.filter(f, queue)}}
+  end
```

Comment:
You are absolutely correct, I have failed to account for any push waiters as my use case was with infinite queues. I have pushed a new commit that takes this into account